res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,741 @@
+        return;
+    }
+    PioneerDDJ400.loopAdjustIn[channel] = !PioneerDDJ400.loopAdjustIn[channel];
+    PioneerDDJ400.loopAdjustOut[channel] = false;
+};
+
+PioneerDDJ400.toggleLoopAdjustOut = function(channel, _control, value, _status, group) {
+    if (value === 0 || engine.getValue(group, "loop_enabled" === 0)) {
+        return;
+    }
+    PioneerDDJ400.loopAdjustOut[channel] = !PioneerDDJ400.loopAdjustOut[channel];
+    PioneerDDJ400.loopAdjustIn[channel] = false;
+};
+
+PioneerDDJ400.setReloopLight = function(status, value) {
+    midi.sendShortMsg(status, 0x4D, value);
+    midi.sendShortMsg(status, 0x50, value);
+};
+
+
+PioneerDDJ400.setLoopButtonLights = function(status, value) {
+    [0x10, 0x11, 0x4E, 0x4C].forEach(function(control) {
+        midi.sendShortMsg(status, control, value);
+    });
+};
+
+PioneerDDJ400.startLoopLightsBlink = function(channel, control, status, group) {
+    var blink = 0x7F;
+
+    PioneerDDJ400.stopLoopLightsBlink(group, control, status);
+
+    PioneerDDJ400.timers[group][control] = engine.beginTimer(500, function() {
+        blink = 0x7F - blink;
+
+        // When adjusting the loop out position, turn the loop in light off
+        if (PioneerDDJ400.loopAdjustOut[channel]) {
+            midi.sendShortMsg(status, 0x10, 0x00);
+            midi.sendShortMsg(status, 0x4C, 0x00);
+        } else {
+            midi.sendShortMsg(status, 0x10, blink);
+            midi.sendShortMsg(status, 0x4C, blink);
+        }
+
+        // When adjusting the loop in position, turn the loop out light off
+        if (PioneerDDJ400.loopAdjustIn[channel]) {
+            midi.sendShortMsg(status, 0x11, 0x00);
+            midi.sendShortMsg(status, 0x4E, 0x00);
+        } else {
+            midi.sendShortMsg(status, 0x11, blink);
+            midi.sendShortMsg(status, 0x4E, blink);
+        }
+    });
+
+};
+
+PioneerDDJ400.stopLoopLightsBlink = function(group, control, status) {
+    PioneerDDJ400.timers[group] = PioneerDDJ400.timers[group] || {};
+
+    if (PioneerDDJ400.timers[group][control] !== undefined) {
+        engine.stopTimer(PioneerDDJ400.timers[group][control]);
+    }
+    PioneerDDJ400.timers[group][control] = undefined;
+    PioneerDDJ400.setLoopButtonLights(status, 0x7F);
+};
+
+PioneerDDJ400.loopToggle = function(value, group, control) {
+    var status = group === "[Channel1]" ? 0x90 : 0x91,
+        channel = group === "[Channel1]" ? 0 : 1;
+
+    PioneerDDJ400.setReloopLight(status, value ? 0x7F : 0x00);
+
+    if (value) {
+        PioneerDDJ400.startLoopLightsBlink(channel, control, status, group);
+    } else {
+        PioneerDDJ400.stopLoopLightsBlink(group, control, status);
+        PioneerDDJ400.loopAdjustOut[channel] = PioneerDDJ400.loopAdjustIn[channel] = false;
+    }
+};
+
+//
+// CUE/LOOP CALL
+//
+
+PioneerDDJ400.cueLoopCallLeft = function(_channel, _control, value, _status, group) {
+    if (value) {
+        engine.setValue(group, "loop_scale", 0.5);
```

Comment:
`loop_scale` is a legacy Control that doesn't change `beatloop_size`. IIRC it was used here because of the manual loop in and out placement buttons. The usual `loop_halve` and `loop_double` change `beatloop_size` without changing the size of the loop if the size of the loop is not a number of beats. So, how about using `loop_scale` if quantize if off, but if quantize is on, use `loop_double`/`loop_halve.

Reply:
:thinking: This opens up a corner case where the controller will only change `beatloop_size` and not the loop as expected if quantize is turned on during a loop (admittedly not a likely thing to do).

Reply:
I understand why it's good to try and keep Mixxx's UI and the controller functions aligned, but in this case I don't think we can do it fully.

I feel we should just leave it as is. Happy to keep the suggested change though if you'd prefer @Be-ing?

Either way, I think we need to add a note to the manual to make whatever decision we make here clear to the user.

Reply:
Hmm, this is tricky. You could change `beatloop_size` directly when quantize is enabled without using `loop_double`/`loop_halve` and always use `loop_scale` unconditionally...

Reply:
The corner case of enabling quantize after a loop is set still creates a problem though. Then the value of `beatloop_size` will be doubled/halved regardless of whether it resembles the actual loop length.

It feels like this is suggesting an issue with the controls. `loop_scale`, `loop_double` and `loop_halve` are leaking their implementation concerns (i.e., `beatloop_size` being equal to the actual loop size) into the script logic.

Possibly worth logging a card/issue (sorry I don't know how the team manages these things) for separate discussion? Happy to look into that separately too (just let me know)

Reply:
We discussed this a lot when implementing the current interface (#1187). Unfortunately, we could not figure out any solution that doesn't leave some awkward corner cases. If you manage to figure out a better solution, that would be great, but I'm not sure that one exists.